extendr-api/src/graphics/mod.rs

```diff
@@ -0,0 +1,759 @@
+
+    // pub fn char_extra_size(&mut self, cex: f64) -> &mut Self {
+    //     self.context.cex = cex;
+    //     self
+    // }
+
+    /// Set the font face.
+    /// ```norun
+    ///   FontFace::PlainFont
+    ///   FontFace::BoldFont
+    ///   FontFace::ItalicFont
+    ///   FontFace::BoldItalicFont
+    ///   FontFace::SymbolFont
+    /// ```
+    pub fn font_face(&mut self, fontface: FontFace) -> &mut Self {
+        use FontFace::*;
+        self.context.fontface = match fontface {
+            PlainFont => 1,
+            BoldFont => 2,
+            ItalicFont => 3,
+            BoldItalicFont => 4,
+            SymbolFont => 5,
+        };
+        self
+    }
+
+    //
+    pub fn font_family(&mut self, fontfamily: &str) -> &mut Self {
+        let maxlen = self.context.fontfamily.len() - 1;
+
+        for c in self.context.fontfamily.iter_mut() {
+            *c = 0;
+        }
+
+        for (i, b) in fontfamily.bytes().enumerate().take(maxlen) {
+            self.context.fontfamily[i] = b as std::os::raw::c_char;
+        }
+        self
+    }
+
+    /// Set the transform as a 3x2 matrix.
+    pub fn transform(
+        &mut self,
+        xscale: (f64, f64),
+        yscale: (f64, f64),
+        offset: (f64, f64),
+    ) -> &mut Self {
+        self.xscale = xscale;
+        self.yscale = yscale;
+        self.offset = offset;
+        self
+    }
+
+    pub(crate) fn context(&self) -> pGEcontext {
+        unsafe { std::mem::transmute(&self.context) }
+    }
+
+    // Affine transform.
+    pub(crate) fn t(&self, xy: (f64, f64)) -> (f64, f64) {
+        (
+            self.offset.0 + xy.0 * self.xscale.0 + xy.1 * self.yscale.0,
+            self.offset.1 + xy.0 * self.xscale.1 + xy.1 * self.yscale.1,
+        )
+    }
+
+    // Affine relative transform (width, height).
+    pub(crate) fn trel(&self, wh: (f64, f64)) -> (f64, f64) {
+        (
+            wh.0 * self.xscale.0 + wh.1 * self.yscale.0,
+            wh.0 * self.xscale.1 + wh.1 * self.yscale.1,
+        )
+    }
+
+    // Scalar transform (eg. radius etc).
+    pub(crate) fn ts(&self, value: f64) -> f64 {
+        value * self.scalar
+    }
+
+    // Inverse scalar transform (eg. text width etc).
+    pub(crate) fn its(&self, value: f64) -> f64 {
+        value / self.scalar
+    }
+
+    pub(crate) fn tmetric(&self, tm: TextMetric) -> TextMetric {
+        TextMetric {
+            ascent: tm.ascent / self.scalar,
+            descent: tm.descent / self.scalar,
+            width: tm.width / self.scalar,
+        }
+    }
+}
+
+#[allow(non_snake_case)]
+impl Device {
+    /// Get the current device.
+    pub fn current() -> Device {
+        unsafe {
+            Device {
+                inner: GEcurrentDevice(),
+            }
+        }
+    }
+
+    /// Enable device rendering.
+    pub fn mode_on(&self) {
+        unsafe {
+            GEMode(1, self.inner());
+        }
+    }
+
+    /// Disable device rendering and flush.
+    pub fn mode_off(&self) {
```

Comment:
I think this fails when the current device is the null device. If so, can we catch it and return `Result`?

Reply:
These two now have an error return in the case of `Rf_NoDevices()`